### ISSUE TITLE
[Bug] wxGUI layout error in GRASS GIS 7.8.dev with Python 3.10

### ISSUE THREAD

**Issue by neteler** [OSGeo/grass#2040]
**Describe the bug**

At time, the wGUI shows an "unusual" layout:

![image](https://user-images.githubusercontent.com/1295172/147421666-680e3f05-eb01-46fd-8b09-45fded6671a6.png)

By clicking into the lower part of the layer manager where the tabs should be all flips into their places.
This shows the error in the console:

```
Traceback (most recent call last):
  File "/usr/lib64/python3.10/site-
packages/wx/lib/agw/flatnotebook.py", line 5205, in OnPaint

renderer.DrawTabs(self, dc)
  File "/usr/lib64/python3.10/site-
packages/wx/lib/agw/flatnotebook.py", line 2360, in DrawTabs

self.DrawTab(pc, dc, posx, i, tabWidth, tabHeight,
pc._nTabXButtonStatus)
  File "/usr/lib64/python3.10/site-
packages/wx/lib/agw/flatnotebook.py", line 2732, in DrawTab

DrawButton(dc, rr, pc.GetSelection() == tabIdx , not
pc.HasAGWFlag(FNB_BOTTOM))
  File "/usr/lib64/python3.10/site-
packages/wx/lib/agw/flatnotebook.py", line 928, in
DrawButton

leftPt = wx.Point(rect.x, rect.y + (rect.height / 10)*8)
TypeError
:
Point(): arguments did not match any overloaded call:
  overload 1: too many arguments
  overload 2: argument 2 has unexpected type 'float'
  overload 3: argument 1 has unexpected type 'int'
  overload 4: argument 1 has unexpected type 'int'
```

**System description (please complete the following information):**

- Operating System: Fedora 35

```
GRASS version: 7.8.7dev                                                         
Code revision: f9f30f33f                                                        
Build date: 2021-12-14                                                          
Build platform: x86_64-pc-linux-gnu                                             
GDAL: 3.3.2                                                                     
PROJ: 8.2.0                                                                     
GEOS: 3.9.2                                                                     
SQLite: 3.36.0                                                                  
Python: 3.10.1                                                                  
wxPython: 4.0.7                                                                 
Platform: Linux-5.15.6-200.fc35.x86_64-x86_64-with-glibc2.34
```

Note: this is wxPython: 4.0.7-23 (i.e., with multiple recent patches to make the wxGUI start).

**Comment by petrasovaa** [comment]
See https://github.com/OSGeo/grass/issues/2019#issuecomment-1001299603

**Comment by neteler** [comment]
(reported to Fedora in https://bugzilla.redhat.com/show_bug.cgi?id=2035790)

**Comment by swt2c** [comment]
Can someone please test if this patch resolves the issue?  The version of GRASS in F35 doesn't seem to manifest the issue - at least I can't reproduce that traceback, or don't know how.

https://github.com/wxWidgets/Phoenix/pull/2062

**Comment by neteler** [comment]
I have tried it but no layout difference yet. 
Now I get this error in the GRASS GIS console:

```
Traceback (most recent call last):
  File "/usr/lib64/python3.10/site-
packages/wx/lib/agw/flatnotebook.py", line 5205, in OnPaint

renderer.DrawTabs(self, dc)
  File "/usr/lib64/python3.10/site-
packages/wx/lib/agw/flatnotebook.py", line 2360, in DrawTabs

self.DrawTab(pc, dc, posx, i, tabWidth, tabHeight,
pc._nTabXButtonStatus)
  File "/usr/lib64/python3.10/site-
packages/wx/lib/agw/flatnotebook.py", line 2772, in DrawTab

dc.DrawText(pc.GetPageText(tabIdx), posx + textOffset,
imageYCoord)
TypeError
:
DC.DrawText(): arguments did not match any overloaded call:
  overload 1: argument 2 has unexpected type 'float'
  overload 2: argument 2 has unexpected type 'float'
```

**Comment by swt2c** [comment]
Thanks - it seems more fixes are needed.  Can you tell me how I can reproduce this myself so I can try to fix them all without having to go back and forth?

**Comment by neteler** [comment]
@swt2c thanks for your efforts! I simply start it with

`grass78 --gui`

Then, if you don't have, get some sample data (note that this is all ways easier in the future GRASS GIS 8 version):

![image](https://user-images.githubusercontent.com/1295172/147501410-aa7c44cf-901e-4e92-98d1-a805000e98de.png)

Next you should see this and may start the session:

![image](https://user-images.githubusercontent.com/1295172/147501807-533e3d11-2ce5-4df1-a470-9b2c040d8bdd.png)

This results into:
![image](https://user-images.githubusercontent.com/1295172/147501871-05621abf-3d08-4f89-abf5-07e2219a02a0.png)

**Comment by swt2c** [comment]
Thanks, I can reproduce now!

**Comment by swt2c** [comment]
Should be fixed by: https://github.com/wxWidgets/Phoenix/commit/aa3dca0e40bd0701e82ce40297a982b5b84844dc

**Comment by neteler** [comment]
Excellent! (I patched wxpython 4.0.7 locally)

![image](https://user-images.githubusercontent.com/1295172/147530617-8815475f-cd17-49be-9ffd-9f655beb675e.png)

@swt2c one question: how to report further wxpython/Python 3.10 issues (I found one, by loading a raster map and switching to 3D mode). I suppose this better done in Fedora bugzilla?

**Comment by swt2c** [comment]
Actually - upstream at https://github.com/wxWidgets/Phoenix/issues is probably the best place.  I'll then backport whatever fixes you need to Fedora.

**Comment by neteler** [comment]
Thanks Scott. I'll try my best but every single GUI function needs to be tested (again) and issues reported. I'll not be able to handle that (hey GRASS users, please assist).

We are super close to release, now I don't know how to handle that... go on and just bad luck for ppl using wxpython/Python 3.10?

**Comment by swt2c** [comment]
> Thanks Scott. I'll try my best but every single GUI function needs to be tested (again) and issues reported. I'll not be able to handle that (hey GRASS users, please assist).

I understand completely.  I went through all the wxPython demos with Python 3.10, but obviously those don't exercise 100% of the code as you're finding.  :-(

> We are super close to release, now I don't know how to handle that... go on and just bad luck for ppl using wxpython/Python 3.10?

Hard to say - maybe recommend Python 3.9 or lower for now?

**Comment by neteler** [comment]
> Hard to say - maybe recommend Python 3.9 or lower for now?

Maybe yes - but can I enforce it e.g. in Fedora 35 (I mean, compatible to non-expert users)?

**Comment by swt2c** [comment]
> Maybe yes - but can I enforce it e.g. in Fedora 35 (I mean, compatible to non-expert users)?

Good point.  There is Python 3.9 in the F35 repository, but not wxPython (and possibly other dependencies).  I don't have a good answer.  :-(

**Comment by neteler** [comment]
> > Thanks Scott. I'll try my best but every single GUI function needs to be tested (again) and issues reported. I'll not be able to handle that (hey GRASS users, please assist).
> 
> I understand completely. I went through all the wxPython demos with Python 3.10, but obviously those don't exercise 100% of the code as you're finding. :-(

(A lot has been fixed in #2050).

This issue has been fixed as well, thanks for the support, @swt2c!
